Adds generic `get`/`get_all` for `Struct` in terms of `AsRef<str>`.

almann opened the pull request:
* Ergonomic improvement of constructing a borrowed/owned `Struct` easier from an interator.
  (see #136).
* Adds a doc test that demonstrates the above constructor and the generic API in use.
* Also did some minor clean-up around some of the `get`/`get_all` implementation to
  be a little less verbose.

Resolves #142.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

almann commented on src/value/borrowed.rs:
```diff
@@ -123,12 +124,32 @@ pub struct BorrowedStruct<'val> {
-    fn new(
-        text_fields: HashMap<String, Vec<(BorrowedSymbolToken<'val>, BorrowedElement<'val>)>>,
-        no_text_fields: Vec<(BorrowedSymbolToken<'val>, BorrowedElement<'val>)>,
-    ) -> Self {
+impl<'val, K, V> FromIterator<(K, V)> for BorrowedStruct<'val>
+where
+    K: Into<BorrowedSymbolToken<'val>>,
+    V: Into<BorrowedElement<'val>>,
+{
+    /// Returns a borrowed struct from the given iterator of field names/values.
+    fn from_iter<I: IntoIterator<Item = (K, V)>>(iter: I) -> Self {
```
Eliminates the old clunky constructor in favor of this API

desaikd replied:
This looks good

almann commented on src/value/borrowed.rs:
```diff
@@ -143,38 +164,35 @@ impl<'val> Struct for BorrowedStruct<'val> {
     ) -> Box<dyn Iterator<Item = (&'a Self::FieldName, &'a Self::Element)> + 'a> {
-        // convert (&k, &v[0..n]) -> (&k, &v[0])
-        // flattens the fields_with_text_key HashMap and chains with fields_with_no_text_key
+        // flattens the text_fields HashMap and chains with no_text_fields
```
Eliminated this comment--didn't seem particularly germane.

almann commented on src/value/borrowed.rs:
```diff
@@ -143,38 +164,35 @@ impl<'val> Struct for BorrowedStruct<'val> {
-            None => None,
-            Some(value) => Some(&value.1),
-        }
+    fn get<T: AsRef<str>>(&self, field_name: T) -> Option<&Self::Element> {
+        self.text_fields
+            .get(field_name.as_ref())?
+            .last()
+            .map(|(_s, v)| v)
```
Made this more concise--feel free to disagree.

desaikd replied:
No! I think this makes sense as per the parameters to the method.

almann commented on src/value/mod.rs:
````diff
@@ -348,13 +348,35 @@ pub trait Struct {
+    /// ```
+    /// # use ion_rs::value::*;
+    /// # use ion_rs::value::owned::*;
+    /// let fields: Vec<(&str, OwnedValue)>= vec![("a", "b"), ("c", "d")]
+    ///     .into_iter().map(|(k, v)| (k, OwnedValue::String(v.into()))).collect();
+    /// let owned: OwnedStruct = fields.into_iter().collect();
+    /// assert_eq!("d", owned.get("c".to_string()).map(|e| e.as_str()).flatten().unwrap());
````
Tests the owned variant with an explicit lookup with a `String`.

almann commented on src/value/mod.rs:
````diff
@@ -348,13 +348,35 @@ pub trait Struct {
+    /// ```
+    /// # use ion_rs::value::*;
+    /// # use ion_rs::value::borrowed::*;
+    /// let fields: Vec<(&str, BorrowedValue)>= vec![("a", "b"), ("c", "d"), ("c", "e")]
+    ///     .into_iter().map(|(k, v)| (k, BorrowedValue::String(v))).collect();
+    /// let owned: BorrowedStruct = fields.into_iter().collect();
+    /// assert_eq!(
+    ///     vec!["d", "e"],
+    ///     owned.get_all("c").flat_map(|e| e.as_str()).collect::<Vec<&str>>()
````
Tests the borrow variant.

almann commented on src/value/owned.rs:
```diff
@@ -158,25 +179,23 @@ impl Struct for OwnedStruct {
 
-    fn get_all<'a>(
+    fn get_all<'a, T: AsRef<str>>(
         &'a self,
-        field_name: &'a str,
+        field_name: T,
```
Note that the lifetime parameter is not needed for `field_name` it gets elided anyhow with this refactor.

desaikd reviewed (approved):
LGTM! Do we want to add doc test for `from_iter` in `Borrowed/OwnedStruct`?

almann commented:
> LGTM! Do we want to add doc test for `from_iter` in `Borrowed/OwnedStruct`?

Good point, I added them, though we probably should factor more comprehensive tests as unit tests and cleanup the doc tests to make them more concise.  I added #145 to track this.

zslayton commented on src/value/borrowed.rs:
```diff
@@ -123,12 +124,32 @@ pub struct BorrowedStruct<'val> {
+{
+    /// Returns a borrowed struct from the given iterator of field names/values.
+    fn from_iter<I: IntoIterator<Item = (K, V)>>(iter: I) -> Self {
+        let mut text_fields: HashMap<String, Vec<(BorrowedSymbolToken, BorrowedElement)>> =
```
Food for thought: we could optimize for the common case by using [`SmallVec`](https://doc.servo.org/smallvec/struct.SmallVec.html) instead of `Vec`.

almann replied:
Added #154 for this.
